README.md
# zlog

<!-- badges: start -->
[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![license](http://img.shields.io/badge/license-GPL%20%28%3E=%203%29-brightgreen.svg?style=flat)](http://www.gnu.org/licenses/gpl-3.0.html)
[![R build status](https://github.com/ampel-leipzig/zlog/workflows/R-CMD-check/badge.svg)](https://github.com/ampel-leipzig/zlog/actions)
[![Codecov test coverage](https://codecov.io/gh/ampel-leipzig/zlog/branch/main/graph/badge.svg)](https://codecov.io/gh/ampel-leipzig/zlog?branch=main)
[![cran checks](https://cranchecks.info/badges/worst/zlog)](https://cran.r-project.org/web/checks/check_results_zlog.html)
[![metacran version](http://www.r-pkg.org/badges/version/zlog)](http://cran.r-project.org/web/packages/zlog/index.html)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.4727117.svg)](https://doi.org/10.5281/zenodo.4727117)
<!-- badges: end -->

Transformation of Laboratory Measurements into z(log)-values

## Description

`zlog` provides functions to transform laboratory measurements into
z(log)-values based on given or empirical reference limits as proposed in
[Hoffmann et al. 2017](https://doi.org/10.1515/labmed-2016-0087).

Basic usage: https://ampel-leipzig.github.io/zlog/articles/zlog.html

## Install

You can install the latest release directly from
[CRAN](http://cran.r-project.org/package=zlog).

```r
install.packages("zlog")
```

## Install development version (not recommended)

[GitHub](https://github.com) is not directly supported by the basic
`install.packages` command. You could use the
[remotes](http://cran.r-project.org/package=remotes) package
to install the development version of `zlog`.

```r
install.packages("remotes")
remotes::install_github("ampel-leipzig/zlog")
```

## Contact

You are welcome to:

* submit suggestions and bug-reports at: <https://github.com/ampel-leipzig/zlog/issues>
* send a pull request on: <https://github.com/ampel-leipzig/zlog/pulls>
* compose an e-mail to: <mail@sebastiangibb.de>


NEWS.md
# zlog 1.0

## Changes 1.0.2 (2022-12-28)

- Ignore cases for the "age" and "sex" column names in
  `lookup_limits` and `.lookup_limits_df` (now case insensitive).
  Affected user visible functions: `zlog_df`, `impute_df`.

## Changes 1.0.1 (2021-05-01)

- Add URL and BugReport URL to DESCRIPTION file.

## Changes in 1.0.0 (2021-04-29)

- First CRAN release.

# zlog 0.0

## Changes in 0.0.16

- Add `impute_df` to impute missing values with there correct reference values.

## Changes in 0.0.15

- No user-visible changes.
- Move reference lookup code from `z_df`/`zlog_df` to the internal
  `.lookup_limits_df` function.

## Changes in 0.0.14

- Fix for `lookup_limits` that reordered its results depending on the "age"
  column in the "table" argument.
- Fix for `z_df` and `z_logdf` now correctly handle the ordering of the
  `matrix` returned by `lookup_limits`.

## Changes in 0.0.13

- Add `set_missing_limits`.
- Fix first author name of reference in manual pages for `zcol`, `zlog` and
  `zlog_df`.

## Changes in 0.0.12

- `z_df` and `zlog_df` now accept a reference table `limits` with more
  parameters (`params`) than present in the `x` `data.frame`.

## Changes in 0.0.11

- Add `z_df` and `zlog_df` as convenient wrapper around `z`/`zlog` and
  `lookup_limits` for larger `data.frame`s and reference tables.
- Fix ordering in `lookup_limits` if a `"param"` column is given.

## Changes in 0.0.10

- `z` gains a `log` argument (default: `FALSE`) and `zlog` is now an alias for
  `z(..., log = TRUE)`.

## Changes in 0.0.9

- Add support for multiple reference parameters in `lookup_limits`.

## Changes in 0.0.8

- Add *z* and *z(log)* equation to vignette.

## Changes in 0.0.7

- Extend `pbc` vignette example with a sex-specific reference table and use
  `lookup_limits`.

## Changes in 0.0.6

- `zcol`: slightly increase color contrast around 1 and 2
  (move darker colours one unit into the middle).

## Changes in 0.0.5

- Add `lookup_limits` to find reference limits in reference tables.

## Changes in 0.0.4

- Fix `zcol` for `NA`.
- Add `surival::pbc` example to the vignette.

## Changes in 0.0.3

- Add `iz`/`izlog` as inverse function to `z`/`zlog`;
  closes [#2](https://github.com/ampel-leipzig/zlog/issues/2).

## Changes in 0.0.2

- Vectorise `limits`, `probs` argument of `z` and `zlog`;
  closes [#1](https://github.com/ampel-leipzig/zlog/issues/1).

## Changes in 0.0.1

- Initial version.


tests/testthat.R
library("testthat")
library("zlog")

test_check("zlog")


tests/testthat/test_zcol.R
test_that("zcol", {
    expect_error(zcol("foo"))
    expect_identical(zcol(c(0, 10)), c("#FEFCFA", "#F57800"))
    expect_identical(zcol(NA), c("#FEFCFA"))
})

test_that(".zlogF", {
    expect_identical(.zlogF(-2, 255), .zlogF(2, 255))
    expect_equal(.zlogF(0, 255), 4.59, tolerance = 0.01)
    expect_equal(.zlogF(11, 255), 255, tolerance = 0.01)
    expect_equal(.zlogF(-11, 120), 120, tolerance = 0.01)
})


tests/testthat/test_zlog.R
test_that("z throws errors", {
    expect_error(z())
    expect_error(z(1:3), "\"limits\" is missing")
    expect_error(z(1:3, limits = "foo"), "numeric")
    expect_error(z(1:3, limits = 1), "length 2")
    expect_error(z(1:3, limits = cbind(1:2, 1:2)), "number of rows")
    expect_error(z(1:3, limits = c(1, 2), probs = "foo"), "numeric")
    expect_error(z(1:3, limits = c(1, 2), probs = 1), "length 2")
    expect_error(z(1:3, limits = cbind(1:3, 1:3), probs = cbind(1:2, 1:2)),
                 "number of rows")
})

test_that("zlog throws errors", {
    expect_error(zlog())
    expect_error(zlog(1:3), "\"limits\" is missing")
    expect_error(zlog(1:3, limits = "foo"), "numeric")
    expect_error(zlog(1:3, limits = 1), "length 2")
    expect_error(zlog(1:3, limits = c(1, 2), probs = "foo"), "numeric")
    expect_error(zlog(1:3, limits = c(1, 2), probs = 1), "length 2")
})

test_that("z/zlog reproduce the results in Hoffmann et al. 2017", {
    # from Hoffmann et al. 2017
    alb <- c(42, 34, 38, 43, 50, 42, 27, 31, 24)
    albz <- c(-0.15, -2.25, -1.15, 0.08, 1.57, -0.15, -4.53, -3.16, -5.70)
    expect_equal(z(log(alb), limits = log(c(35, 52))), albz, tolerance = 0.002)
    expect_equal(zlog(alb, limits = c(35, 52)), albz, tolerance = 0.002)

    bili <- c(11, 9, 2, 5, 22, 42, 37, 200, 20)
    biliz <- c(0.85, 0.57,
               -1.96,
               # the zlog in Hoffmann et al. 2017 is -1.66 but seems to be wrong
               # cause the value 2 is exactly the lower limit which should be
               # the lower limit (-1.96)
               -0.43, 2.04, 3.12, 2.90, 5.72, 1.88)
    nalb <- length(alb)
    nbili <- length(bili)
    nn <- c(nalb, nbili)
    limits <- cbind(ll = rep(c(35, 2), nn), ul = rep(c(52, 21), nn))
    expect_equal(
        zlog(c(alb, bili), limits = limits), c(albz, biliz), tolerance = 0.005
    )
})

test_that("z/zlog handles a matrix as probs", {
    expect_equal(
        zlog(c(2, 5),
          limits = cbind(rep(2, 2), rep(5, 2)),
          probs = cbind(c(0.1, 0.2), c(0.9, 0.8))
        ),
        qnorm(c(0.1, 0.8))
    )
})


R/zcol.R
#' Z(log) Depending Color
#'
#' This function provides a color gradient depending on the zlog value as
#' described in Hoffmann et al. 2017. The colours are only roughly equal to the
#' one found in the article.
#'
#' @param x `numeric`, z(log) value.
#' @return `character`, of `length(x)` with the corresponding color hex code.
#'
#' @author Sebastian Gibb
#' @references
#' Hoffmann, Georg, Frank Klawonn, Ralf Lichtinghagen, and Matthias Orth.
#' 2017.
#' "The Zlog-Value as Basis for the Standardization of Laboratory Results."
#' LaboratoriumsMedizin 41 (1): 23–32. \doi{10.1515/labmed-2016-0087}.
#' @importFrom graphics image text
#' @importFrom grDevices rgb
#' @export
#' @examples
#' z <- -10:10
#' image(matrix(z, ncol = 1), col = zcol(z), xaxt = "n", yaxt = "n")
#' text(seq(0, 1, length.out=length(z)), 0, label = z)
zcol <- function(x) {
    x[is.na(x)] <- 0

    ## the colour values are picked from the PDF version of Hoffmann et al. 2017
    ## and may be incorrect cause of jpeg artefacts
    rgb(
        red = 255 - .zlogF(x, ifelse(x < 0, 185, 10)),
        green = 255 - .zlogF(x, 135),
        blue = 255 - .zlogF(x, ifelse(x < 0, 70, 255)),
        maxColorValue = 255
    )
}

#' Map zlog Value to 0:255
#'
#' Similar to the `F` described in Hoffmann et al. 2017.
#'
#' @param x `numeric`, zlog value.
#' @param d `numeric`, scale value.
#'
#' @return `numeric`, of `length(x)` in the range `0:255`.
#' @noRd
.zlogF <- function(x, d) {
    # the proposed function in Hoffmann et al. 2017 doesn't increase enough,
    # so we add 4 here
    # the color range is white:color for 0:10, 0:1 should be white, 2 should
    # have a clear visible color, d / 1 + exp(4 - abs(x)) is in the range from
    # 4.6:254.4 for x = 0:10 and d = 255
    d / (1 + exp(4 - abs(x)))
}


R/zlog.R
#' Calculate z/zlog Values
#'
#' Calculates the z or z(log) values for laboratory measurement standardisation
#' as proposed in Hoffmann 2017 et al.
#'
#' @param x `numeric`, laboratory values.
#' @param limits `numeric` or `matrix`, lower and upper reference limits. Has to
#' be of length 2 for `numeric` or a two-column `matrix` with as many rows as
#' elements in `x`.
#' @param probs `numeric`, probabilities of the lower and upper reference limit,
#' default: `c(0.025, 0.975)` (spanning 95 %). Has to be of length 2 for
#' `numeric` or a two-column `matrix` with as many rows as elements in `x`.
#' @param log `logical`, should z (`log = FALSE`, default) or
#' z(log) (`log = TRUE`) calculated?
#'
#' @details
#' The z value is calculated as follows (assuming that the limits where 0.025
#' and 0.975 quantiles):
#' \eqn{z = (x - (limits_1 + limits_2 )/2) * 3.92/(limits_2 - limits_1)}.
#'
#' The z(log) value is calculated as follows (assuming that the limits where 0.025
#' and 0.975 quantiles):
#' \eqn{z = (\log(x) - (\log(limits_1) + \log(limits_2))/2) * 3.92/(\log(limits_2) - \log(limits_1))}.
#'
#' `zlog` is an alias for `z(..., log = TRUE)`.
#'
#' @return `numeric`, z or z(log) values.
#' @rdname zlog
#' @author Sebastian Gibb
#' @references
#' Georg Hoffmann, Frank Klawonn, Ralf Lichtinghagen, and Matthias Orth.
#' 2017.
#' "The Zlog-Value as Basis for the Standardization of Laboratory Results."
#' LaboratoriumsMedizin 41 (1): 23–32.
#' \doi{10.1515/labmed-2016-0087}.
#' @seealso [`izlog()`]
#'
#' @importFrom stats qnorm
#' @export
#' @examples
#' z(1:10, limits = c(2, 8))
#'
z <- function(x, limits, probs = c(0.025, 0.975), log = FALSE) {
    if (!(is.numeric(limits) && length(limits) == 2L) &&
        !(is.matrix(limits) && mode(limits) == "numeric" &&
          nrow(limits) == length(x) && ncol(limits) == 2L))
        stop("'limits' has to be a numeric of length 2, or a ",
             "matrix with 2 columns (lower and upper limit) where ",
             "the number of rows equals the length of 'x'.")

    if (!is.matrix(limits))
        limits <- t(limits)

    if (!(is.numeric(probs) && length(probs) == 2L) &&
        !(is.matrix(probs) && mode(probs) == "numeric" &&
          nrow(probs) == length(x) && ncol(probs) == 2L))
        stop("'probs' has to be a numeric of length 2, or a ",
             "matrix with 2 columns (lower and upper limit) where ",
             "the number of rows equals the length of 'x'.")

    if (!is.matrix(probs))
        probs <- t(probs)

    if (log) {
        x <- log(x)
        limits <- log(limits)
    }

    m <- (limits[, 1L] + limits[, 2L]) / 2L
    s <- abs(limits[, 2L] - limits[, 1L]) / rowSums(abs(qnorm(probs)))

    (x - m) / s
}

#' @rdname zlog
#' @export
#' @examples
#' # from Hoffmann et al. 2017
#' albumin <- c(42, 34, 38, 43, 50, 42, 27, 31, 24)
#' zlog(albumin, limits = c(35, 52))
#'
#' bilirubin <- c(11, 9, 2, 5, 22, 42, 37, 200, 20)
#'
#' limits <- cbind(
#'     lower = rep(c(35, 2), c(length(albumin), length(bilirubin))),
#'     upper = rep(c(52, 21), c(length(albumin), length(bilirubin)))
#' )
#' zlog(c(albumin, bilirubin), limits = limits)
zlog <- function(x, limits, probs = c(0.025, 0.975)) {
    z(x, limits, probs, log = TRUE)
}


R/impute.R
#' Imputation
#'
#' Impute `NA` values with the logmean, mean, minimal or maximum reference value.
#'
#' @param x `data.frame`, with the columns: "age", `numeric`, "sex", `factor`
#' and more user defined `numeric` columns that should be imputed.
#' @param limits `data.frame`, reference table, has to have the columns:
#' "age", `numeric` (same units as in `age`, e.g. days or years, age of `0`
#' matches all ages),
#' "sex", `factor` (same levels for male and female as `sex` and a special level
#' `"both"`), "param", `character` with the laboratory parameter name that have
#' to match the column name in `x`, "lower" and "upper", `numeric` for the
#' lower and upper reference limits.
#' @param method `character`, imputation method. `method = "logmean"` (default)
#' replaces all `NA` with its corresponding logged mean values for the reference
#' table `limits` (for subsequent use of the *zlog* score,
#' use `method = "mean" for *z* score calculation).
#' For `method = "min"` or `method = "max"` the lower or the upper limits are
#' used.
#'
#' @note
#' Imputation should be done prior to [`z()`]/[`zlog()`] transformation.
#' Afterwards the `NA` could replaced by zero (for mean-imputation) via
#' `d[is.na(d)] <- 0`.
#' @return `data.frame`, the same as `x` but missing values are replaced by
#' the corresponding logmean, mean, minimal or maximal reference values
#' depending on the chosen `method`.
#'
#' @author Sebastian Gibb
#' @export
#' @examples
#' l <- data.frame(
#'     param = c("alb", "bili"),
#'     age = c(0, 0),
#'     sex = c("both", "both"),
#'     units = c("mg/l", "µmol/l"),
#'     lower = c(35, 2),
#'     upper = c(52, 21)
#' )
#' x <- data.frame(
#'     age = 40:48,
#'     sex = rep(c("female", "male"), c(5, 4)),
#'     # from Hoffmann et al. 2017
#'     alb = c(42, NA, 38, NA, 50, 42, 27, 31, 24),
#'     bili = c(11, 9, NA, NA, 22, 42, NA, 200, 20)
#' )
#' impute_df(x, l)
#' impute_df(x, l, method = "min")
#' zlog_df(impute_df(x, l), l)
impute_df <- function(x, limits, method = c("logmean", "mean", "min", "max")) {
    limits <- .lookup_limits_df(x, limits)
    prms <- unique(rownames(limits))

    v <- switch(
        match.arg(method),
        "min" = limits[, "lower"],
        "max" = limits[, "upper"],
        "mean" = (limits[, "lower"] + limits[, "upper"]) / 2L,
        "logmean" = exp(log(limits[, "lower"] * limits[, "upper"]) / 2L)
    )
    na <- is.na(x[prms])
    x[prms][na] <- v[na]
    x
}


R/zlog_df.R
#' Calculate z/zlog Values for a data.frame
#'
#' Calculates the z or z(log) values for laboratory measurement standardisation
#' as proposed in Hoffmann 2017 et al. for a complete `data.frame`.
#'
#' @param x `data.frame`, with the columns: "age", `numeric`, "sex", `factor`
#' and more user defined `numeric` columns that should be z/z(log) transformed.
#' @param limits `data.frame`, reference table, has to have the columns:
#' "age", `numeric` (same units as in `age`, e.g. days or years, age of `0`
#' matches all ages),
#' "sex", `factor` (same levels for male and female as `sex` and a special level
#' `"both"`), "param", `character` with the laboratory parameter name that have
#' to match the column name in `x`, "lower" and "upper", `numeric` for the
#' lower and upper reference limits.
#' @param probs `numeric`, probabilities of the lower and upper reference limit,
#' default: `c(0.025, 0.975)` (spanning 95 %). Has to be of length 2 for
#' `numeric` or a two-column `matrix` with as many rows as elements in `x`.
#' @param log `logical`, should z (`log = FALSE`, default) or
#' z(log) (`log = TRUE`) calculated?
#'
#' @details
#' This is a wrapper function for [`z()`] and [`lookup_limits()`]. Please find
#' the details for the z/z(log) calculation at [`z()`].
#'
#' `zlog_df` is an alias for `z_df(..., log = TRUE)`.
#'
#' @return `data.frame`, with the columns: "age", "sex" and all `numeric`
#' columns z/zlog transformed. If a column name is missing in `limits$param`
#' a warning is thrown and the column is set to `NA`.
#' @rdname zlog_df
#' @author Sebastian Gibb
#' @references
#' Georg Hoffmann, Frank Klawonn, Ralf Lichtinghagen, and Matthias Orth.
#' 2017.
#' "The Zlog-Value as Basis for the Standardization of Laboratory Results."
#' LaboratoriumsMedizin 41 (1): 23–32.
#' \doi{10.1515/labmed-2016-0087}.
#' @seealso [`zlog()`]
#'
#' @export
#' @examples
#' l <- data.frame(
#'     param = c("alb", "bili"),
#'     age = c(0, 0),
#'     sex = c("both", "both"),
#'     units = c("mg/l", "µmol/l"),
#'     lower = c(35, 2),
#'     upper = c(52, 21)
#' )
#' x <- data.frame(
#'     age = 40:48,
#'     sex = rep(c("female", "male"), c(5, 4)),
#'     # from Hoffmann et al. 2017
#'     alb = c(42, 34, 38, 43, 50, 42, 27, 31, 24),
#'     bili = c(11, 9, 2, 5, 22, 42, 37, 200, 20)
#' )
#' z_df(x, l)
#'
z_df <- function(x, limits, probs = c(0.025, 0.975), log = FALSE) {
    limits <- .lookup_limits_df(x, limits)
    prms <- unique(rownames(limits))

    x[prms] <- z(as.matrix(x[prms]), limits, probs = probs, log = log)
    x
}

#' @rdname zlog_df
#' @export
#' @examples
#' zlog_df(x, l)
zlog_df <- function(x, limits, probs = c(0.025, 0.975)) {
    z_df(x, limits, probs, log = TRUE)
}


R/lookup_limits.R
#' Lookup Limits in Reference Tables
#'
#' Reference limits are specific for age and sex. Each laboratory institute has
#' its own reference limits. This function is useful to query a dataset against
#' this database.
#'
#' @param age `numeric`, patient age.
#' @param sex `character`/`factor`, patient sex, has to be the same length as
#' `age`.
#' @param table `data.frame`, reference table, has to have the columns:
#' "age", `numeric` (same units as in `age`, e.g. days or years, age of `0`
#' matches all ages),
#' "sex", `factor` (same levels for male and female as `sex` and a special level
#' `"both"`), "lower" and "upper", `numeric` for the lower and upper reference
#' limits. If an "param" column is given the "lower" and "upper" limits for all
#' different values in "param" is returned. Additional columns are allowed (and
#' ignored).
#'
#' @return `matrix`, with 2 columns ("lower", "upper") and as many rows as
#' `length(age)`.
#' @author Sebastian Gibb
#' @importFrom stats setNames quantile
#' @export
#' @examples
#' reference <- data.frame(
#'     param = c("albumin", rep("bilirubin", 4)),
#'     age = c(0, 1, 2, 3, 7),     # days
#'     sex = "both",
#'     units = c("g/l", rep("µmol/l", 4)), # ignored
#'     lower = c(35, rep(NA, 4)),  # no real reference values
#'     upper = c(52, 5, 8, 13, 18) # no real reference values
#' )
#'
#' # lookup albumin reference values for 18 year old woman
#' lookup_limits(
#'     age = 18 * 365.25,
#'     sex = "female",
#'     table = reference[reference$param == "albumin",]
#' )
#'
#' # lookup albumin and bilirubin values for 18 year old woman
#' lookup_limits(
#'     age = 18 * 365.25,
#'     sex = "female",
#'     table = reference
#' )
#'
#' # lookup bilirubin referenc values for infants
#' lookup_limits(
#'     age = 0:8,
#'     sex = rep(c("female", "male"), 5:4),
#'     table = reference[reference$param == "bilirubin",]
#' )
lookup_limits <- function(age, sex, table) {
    if (!is.numeric(age))
        stop("'age' has to be a numeric value.")
    if (length(age) != length(sex))
        stop("'age' and 'sex' have to be of the same length.")

    sex <- as.factor(sex)
    if (nlevels(sex) > 2)
        stop("'sex' has to be a factor of at most 2 levels ",
             "(male, female).")

    cn <- colnames(table) <- tolower(colnames(table))
    if (!all(c("age", "sex", "lower", "upper") %in% cn))
        stop("'table' has to have the columns: ",
             "\"age\", \"sex\", \"upper\", \"lower\".")

    table$sex <- as.factor(table$sex)
    if (nlevels(table$sex) > 3)
        stop("'table$sex' has to be a factor of at most 3 levels ",
             "(male, female, both).")

    if (!"param" %in% cn)
        table$param <- "param"

    params <- unique(table$param)
    nparam <- length(params)
    nage <- length(age)

    table <- table[order(table$age, decreasing = TRUE),]

    limits <- matrix(
        NA_real_, nrow = nage * nparam, ncol = 2L,
        dimnames = list(
                if (nparam > 1L || params != "param")
                    rep(params, each = nage)
                else
                    NULL,
                c("lower", "upper")
        )
    )

    for (i in seq(along = params)) {
        for (j in seq(along = age)) {
            k <- which(
                (sex[j] == table$sex | table$sex == "both") &
                age[j] >= table$age & params[i] == table$param
            )[1L]
            if (length(k))
                limits[(i - 1) * nage + j, ] <-
                    c(table$lower[k], table$upper[k])
        }
    }

    limits
}

#' Lookup Limits in Reference Tables for a Whole data.frame
#'
#' internal function
#'
#' @param x `data.frame`
#' @param table `data.frame`, reference table, format see above
#' @return see above
#'
#' @noRd
.lookup_limits_df <- function(x, limits) {
    if (!is.data.frame(x)) {
        stop("'x' has to be a data.frame.")
    }

    cnx <- colnames(x)
    cnxl <- tolower(cnx)

    if (!"age" %in% cnxl)
        stop("Column \"age\" is missing in 'x'.")
    if (!"sex" %in% cnxl)
        stop("Column \"sex\" is missing in 'x'.")


    if (!is.data.frame(limits))
        stop("'limits' has to be a data.frame.")

    cnl <- colnames(limits) <- tolower(colnames(limits))

    if (!all(c("age", "sex", "param", "lower", "upper") %in% cnl))
        stop("'limits' has to be a data.frame with the following columns: ",
             "\"age\", \"sex\", \"param\", \"upper\", \"lower\".\n",
             "See '?lookup_limits' for details.")

    num <- vapply(x, is.numeric, FALSE, USE.NAMES = FALSE) &
        !cnxl %in% c("age", "sex")

    if (any(!cnx[num] %in% limits$param)) {
        na <- cnx[num & (!cnx %in% limits$param)]
        warning(
            "No reference for column(s): ", paste0(na, collapse = ", ")
        )
        ina <- nrow(limits) + seq_len(length(na))
        limits[ina, c("param", "age", "sex")] <-
            cbind.data.frame(param = na, age = 0L, sex = "both")
    }

    lookup_limits(
        age = x[[which(cnxl == "age")]], sex = x[[which(cnxl == "sex")]],
        table = limits[limits$param %in% cnx[num],]
    )
}


R/reference_limits.R
#' Calculate Reference Limits
#'
#' Calculates the lower and upper reference limit for given probabilities.
#'
#' @param x `numeric`, laboratory values
#' @param probs `numeric`, probabilities of the lower and upper reference limit,
#' default: `c(0.025, 0.975)` (spanning 95 %).
#' @param na.rm `logical`, if `TRUE` (default) `NA` values are removed before
#' the reference limits are calculated.
#'
#' @return `numeric` of length 2 with the lower and upper limit.
#'
#' @author Sebastian Gibb
#' @importFrom stats setNames quantile
#' @export
#' @examples
#' reference_limits(1:10)
reference_limits <- function(x, probs = c(0.025, 0.975), na.rm = TRUE) {
    if (!is.numeric(probs) || length(probs) != 2L)
        stop("'probs' has to be a numeric of length 2.")

    setNames(
        quantile(x, probs = range(probs), na.rm = na.rm, names = FALSE),
        c("lower", "upper")
    )
}


R/set_missing_limits.R
#' Set Missing Limits in Reference Tables
#'
#' Sometimes reference limits are not specified. That is often the case for
#' biomarkers that are related to infection or cancer. Using zero as lower
#' boundary results in skewed distributions (Hoffmann et al. 2017; fig. 7).
#' Haeckel et al. 2015 suggested to set the lower reference limit to 0.15 of
#' the upper one.
#'
#' @param x `data.frame`, reference table, has to have the columns:
#' "lower" and "upper", `numeric` for the lower and upper reference
#' limits.  Additional columns are allowed (and ignored).
#' @param fraction `numeric(2)`, targeted fraction of the lower to the upper and
#' the upper to the lower limit. Haeckel et al. 2015 suggested to set the lower
#' limit to 0.15 of the upper one. We choose 20/3 (the reciprocal of 0.15) for
#' the upper to the lower one.
#'
#' @return `data.frame`, the same as `x` but the "lower" and "upper" columns are
#' modified if there were `NA` before.
#' @author Sebastian Gibb
#' @references
#' Georg Hoffmann, Frank Klawonn, Ralf Lichtinghagen, and Matthias Orth.
#' 2017.
#' "The Zlog-Value as Basis for the Standardization of Laboratory Results."
#' LaboratoriumsMedizin 41 (1): 23–32.
#' \doi{10.1515/labmed-2016-0087}.
#'
#' Rainer Haeckel, Werner Wosniok, Ebrhard Gurr and Burkhard Peil.
#' 2015.
#' "Permissible limits for uncertainty of measurement in laboratory medicine"
#' Clinical Chemistry and Laboratory Medicine 53 (8): 1161-1171.
#' \doi{10.1515/cclm-2014-0874}.
#' @export
#' @examples
#' reference <- data.frame(
#'     param = c("albumin", rep("bilirubin", 4)),
#'     age = c(0, 1, 2, 3, 7),             # ignored
#'     sex = "both",                       # ignored
#'     units = c("g/l", rep("µmol/l", 4)), # ignored
#'     lower = c(35, rep(NA, 4)),  # no real reference values
#'     upper = c(52, 5, 8, 13, 18) # no real reference values
#' )
#' set_missing_limits(reference)
#' set_missing_limits(reference, fraction = c(0.2, 5))
set_missing_limits <- function(x, fraction = c(0.15, 20/3)) {
    if (!is.data.frame(x))
        stop("'x' has to be a data.frame")

    if (!all(c("lower", "upper") %in% colnames(x)))
        stop("'x' has to have the columns: \"upper\", \"lower\".")

    if (!is.numeric(fraction) || length(fraction) != 2L)
        stop("'fraction' has to be a numeric of length 2")

    na <- is.na(x[c("lower", "upper")])
    x$lower[na[, "lower"]] <- x$upper[na[, "lower"]] * fraction[1L]
    x$upper[na[, "upper"]] <- x$lower[na[, "upper"]] * fraction[2L]
    x
}


R/izlog.R
#' Calculate Laboratory Measurements from z/zlog Values
#'
#' Inverse function to z or z(log) for laboratory measurement standardisation
#' as proposed in Hoffmann 2017 et al.
#'
#' @param x `numeric`, z/zlog values.
#' @param limits `numeric` or `matrix`, lower and upper reference limits. Has to
#' be of length 2 for `numeric` or a two-column `matrix` with as many rows as
#' elements in `x`.
#' @param probs `numeric`, probabilities of the lower and upper reference limit,
#' default: `c(0.025, 0.975)` (spanning 95 %). Has to be of length 2 for
#' `numeric` or a two-column `matrix` with as many rows as elements in `x`.
#'
#' @details
#' The inverse z value is calculated as follows (assuming that the limits where
#' 0.025 and 0.975 quantiles):
#' \eqn{x = z * (limits_2 - limits_1)/3.92 + (limits_1 + limits_2)/2}
#'
#' The inverse z(log) value is calculated as follows (assuming that the limits
#' where 0.025 and 0.975 quantiles):
#' \eqn{x = \exp(z * (\log(limits_2) - \log(limits_1))/3.92 + (\log(limits_1) + \log(limits_2))/2)}
#'
#' @return `numeric`, laboratory measurements.
#' @rdname izlog
#' @author Sebastian Gibb
#' @references
#' Georg Hoffmann, Frank Klawonn, Ralf Lichtinghagen, and Matthias Orth.
#' 2017.
#' "The Zlog-Value as Basis for the Standardization of Laboratory Results."
#' LaboratoriumsMedizin 41 (1): 23–32.
#' \doi{10.1515/labmed-2016-0087}.
#' @seealso [`zlog()`]
#'
#' @importFrom stats qnorm
#' @export
#' @examples
#' iz(z(1:10, limits = c(2, 8)), limits = c(2, 8))
#'
iz <- function(x, limits, probs = c(0.025, 0.975)) {
    if (!(is.numeric(limits) && length(limits) == 2L) &&
        !(is.matrix(limits) && mode(limits) == "numeric" &&
          nrow(limits) == length(x) && ncol(limits) == 2L))
        stop("'limits' has to be a numeric of length 2, or a ",
             "matrix with 2 columns (lower and upper limit) where ",
             "the number of rows equals the length of 'x'.")

    if (!is.matrix(limits))
        limits <- t(limits)

    if (!(is.numeric(probs) && length(probs) == 2L) &&
        !(is.matrix(probs) && mode(probs) == "numeric" &&
          nrow(probs) == length(x) && ncol(probs) == 2L))
        stop("'probs' has to be a numeric of length 2, or a ",
             "matrix with 2 columns (lower and upper limit) where ",
             "the number of rows equals the length of 'x'.")

    if (!is.matrix(probs))
        probs <- t(probs)

    m <- (limits[, 1L] + limits[, 2L]) / 2L
    s <- abs(limits[, 2L] - limits[, 1L]) / rowSums(abs(qnorm(probs)))
    x * s + m
}

#' @rdname izlog
#' @export
#' @examples
#' # from Hoffmann et al. 2017
#' albuminzlog <- c(-0.15, -2.25, -1.15, 0.08, 1.57, -0.15, -4.53, -3.16, -5.70)
#' izlog(albuminzlog, limits = c(35, 52))
#'
#' bilirubinzlog <- c(0.85, 0.57, -1.96, -0.43, 2.04, 3.12, 2.90, 5.72, 1.88)
#'
#' limits <- cbind(
#'     lower = rep(c(35, 2), c(length(albuminzlog), length(bilirubinzlog))),
#'     upper = rep(c(52, 21), c(length(albuminzlog), length(bilirubinzlog)))
#' )
#' izlog(c(albuminzlog, bilirubinzlog), limits = limits)
izlog <- function(x, limits, probs = c(0.025, 0.975)) {
    if (missing(limits))
        stop("argument \"limits\" is missing, with no default")
    exp(iz(x, log(limits), probs))
}
